### ISSUE TITLE
feat(torch): support configurable activation checkpointing for user-defined layers

### ISSUE THREAD

**Issue by f6ra07nk14** [f6ra07nk14/structcast-model#20]
## Context

`BaseModelBuilder` currently parses every model configuration with the framework-neutral `TemplateLayer`, whose
target is `UserDefinedLayer`. `TorchBuilder` changes the generated intermediate type, but there is no template seam
for Torch-only model fields. Adding activation checkpointing directly to the shared schema would leak Torch concepts
into Keras and Flax; overriding the full Torch builder pipeline would duplicate the existing template, nested-layer,
and flow construction logic.

PyTorch activation checkpointing (`torch.utils.checkpoint.checkpoint`, also commonly called gradient checkpointing)
trades additional recomputation for lower retained activation memory and lower peak VRAM during training. It is
unrelated to the repository's Training state / distributed checkpoint persistence, and it should not be presented as
a general training-time speedup.

Users should be able to enable checkpointing on either the root or a named `UserDefinedLayer` in a Torch DSL
configuration and configure options such as `use_reentrant` and `early_stop`.

## Goal

Add configurable, Torch-only activation checkpointing while introducing a small framework-specific template seam
that preserves the shared builder implementation and can later host Keras- or Flax-specific model options.

The intended DSL shape is:

```yaml
GRADIENT_CHECKPOINTING:
  use_reentrant: false
  early_stop: true
  preserve_rng_state: true
  determinism_check: default
  debug: false
```

`GRADIENT_CHECKPOINTING: true` should enable project defaults, while `false` or an omitted field disables the feature.
Native PyTorch option names should remain lower snake case inside the mapping.

### Framework-specific template seam

- Add `TorchUserDefinedLayer(UserDefinedLayer)` for Torch-only model fields, including `GRADIENT_CHECKPOINTING`.
- Add `TorchTemplateLayer(TemplateLayer)` with `target_type = TorchUserDefinedLayer`.
- Add a class-level template selection seam to `BaseModelBuilder`, defaulting to `TemplateLayer`; `__post_init__`
  must validate through that selected type instead of hard-coding `TemplateLayer`.
- Configure `TorchBuilder` to use `TorchTemplateLayer`. Base, Keras, and Flax builders must continue to use the shared
  `TemplateLayer` unless they later introduce their own framework-specific template types.
- Preserve the existing recursive `type(self)` path so root templates, named `TYPE` layers, and external `CFG` layers
  all receive the Torch schema without a parallel resolution pipeline.
- Add one narrow builder hook (or an equivalently small seam) for propagating framework-specific fields into the
  framework-specific layer intermediate. Do not copy or override the complete `BaseModelBuilder.__call__` flow.
- Keep common flow validation and rendering behavior in `UserDefinedLayer` / `TemplateLayer`; Torch subclasses should
  contain only the delta required by Torch.
- Store the resolved checkpoint configuration on `TorchLayerIntermediate` so generated-class hashing and
  deduplication distinguish otherwise identical layers with different checkpoint behavior.

Making the checkpoint configuration part of `TorchUserDefinedLayer` is important: it keeps the field in the
template's `raw` data, so normal `PARAMETERS`, `_jinja_yaml_`, root/named-layer separation, and validation continue to
work through the existing template implementation.

### Configuration contract

- `use_reentrant` must always be explicit in generated PyTorch calls; the project default should be `false`, matching
  PyTorch's recommendation and supporting nested inputs/outputs and keyword arguments.
- `early_stop` must be configurable and default to PyTorch's normal enabled behavior for non-reentrant checkpointing.
- The implementation should validate the supported stable checkpoint options rather than accept arbitrary keys that
  fail later in generated code. At minimum, support `use_reentrant` and `early_stop`; resolve the supported contract
  for `preserve_rng_state`, `context_fn`, `determinism_check`, and `debug` against the project's PyTorch version range.
- Combinations unsupported by the selected mode must fail clearly during configuration validation where practical.
  In particular, `context_fn`, `determinism_check`, and `debug` are non-reentrant-only, while reentrant checkpointing
  has additional restrictions around nested structures, detached tensors, and required gradients.
- Reserve and document the selected top-level DSL field so it cannot be confused with a named user-defined layer.

### Generated Torch behavior

- The checkpoint boundary is the complete training forward implementation of the marked generated
  `UserDefinedLayer`. A private `_forward_impl`-style function or equivalent may hold the existing flow.
- Invoke checkpointing only while the generated module is training and gradients are enabled. Evaluation,
  inference-only flow, and `torch.no_grad()` execution must call the original implementation directly.
- Each root or nested user-defined layer controls its own checkpoint setting independently. Do not automatically
  checkpoint every descendant when only a parent is marked.
- If both a parent and descendant are explicitly marked, preserve that explicit nesting but document the compounded
  recomputation cost.
- Enabling checkpointing must not introduce a wrapper module or change `named_modules()` paths, parameter names, or
  `state_dict()` keys.
- Disabled checkpointing should preserve existing generated behavior and avoid unnecessary checkpoint imports or
  runtime branches.
- Structured outputs and multiple inputs must work with the default non-reentrant mode.

### PyTorch compatibility

The package currently declares `torch>=2.0.0`, while the linked PyTorch 2.13 interface exposes options that may not be
available with the same call shape throughout that range. The implementation must verify the supported version matrix
and choose one explicit policy:

1. normalize version differences behind a Torch-only implementation seam (for example, map `early_stop` through the
   supported context manager where required),
2. reject unavailable options with a clear version-specific error, or
3. deliberately raise the minimum supported PyTorch version as part of this change.

Generated models must not blindly emit a PyTorch 2.13-only call while the package still advertises compatibility with
older releases.

## Acceptance evidence

- [ ] `TorchBuilder.template` is a `TorchTemplateLayer`; Base, Keras, and Flax builders retain their existing template
      behavior.
- [ ] A root layer and a nested `TYPE`/`CFG` layer can each enable checkpointing and receive independent options.
- [ ] Template parameters and `_jinja_yaml_` can produce the Torch-specific checkpoint configuration through the
      normal template path.
- [ ] Disabled checkpointing generates and executes the existing forward behavior without Torch checkpoint code.
- [ ] Enabled checkpointing recomputes the marked forward region during backward and bypasses checkpointing during
      evaluation or no-grad execution.
- [ ] Checkpointed and non-checkpointed models with identical weights and deterministic inputs produce equivalent
      outputs, input gradients, and parameter gradients.
- [ ] `use_reentrant=false` is covered with structured output and multiple-input cases; `use_reentrant=true` is
      covered for a supported tensor-only case.
- [ ] `early_stop` and the other supported options are verified as passed with the requested values; incompatible
      combinations fail loudly.
- [ ] Dropout or another RNG-using layer verifies the configured RNG-preservation behavior.
- [ ] Existing generated model class names, `named_modules()` paths, weight mapping, and `state_dict()` keys remain
      unchanged.
- [ ] Otherwise identical flows with different checkpoint settings are not incorrectly deduplicated into one
      generated class.
- [ ] Existing Keras and Flax builder tests continue to pass without Torch imports or Torch-only schema fields.
- [ ] The supported PyTorch version policy is documented and tested at the narrowest practical seam.
- [ ] At least one existing `torch.compile` model-generation path has a focused smoke test with the supported default
      checkpoint mode, or an incompatibility is rejected and documented explicitly.
- [ ] A representative CUDA measurement records lower peak allocated memory for an enabled checkpointed layer and
      reports the corresponding step-time/recomputation trade-off. This evidence may be non-CI if GPU CI is not
      available.
- [ ] DSL documentation includes the configuration example, placement rules, defaults, mode restrictions, and the
      compute-for-memory trade-off, including warnings for stateful or side-effecting forward code that will run
      again during backward recomputation.

## Out of scope

- Implementing Keras or Flax rematerialization in this issue. Future framework-specific support should reuse the same
  template selection seam rather than adding fields to the shared schema.
- Automatic selection or partitioning of checkpoint regions; users explicitly mark generated user-defined layers.
- Changing Training state, distributed checkpoint persistence, or resume behavior.
- Integrating Hugging Face model classes' own `supports_gradient_checkpointing` lifecycle.
- Guaranteeing faster training; the intended benefit is lower activation memory at the cost of recomputation.

## References

- [PyTorch 2.13 activation checkpointing](https://docs.pytorch.org/docs/2.13/checkpoint.html)
- [Hugging Face: gradient checkpointing performance trade-off](https://huggingface.co/docs/transformers/v4.26.1/perf_train_gpu_one)
- [Hugging Face modeling utilities](https://huggingface.co/docs/transformers/internal/modeling_utils)
- `src/structcast_model/builders/schema.py`: `UserDefinedLayer`, `Template`, and `TemplateLayer`
- `src/structcast_model/builders/base_builder.py`: `BaseModelBuilder` template parsing and nested-layer resolution
- `src/structcast_model/builders/torch_builder.py`: `TorchBuilder` and `TorchLayerIntermediate`
- ADR-0004: generated code owns training-flow behavior and compile boundaries
